File: app/src/main/java/com/shakil/barivara/activities/room/RoomListActivity.java
package com.shakil.barivara.activities.room;

import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.shakil.barivara.R;
import com.shakil.barivara.activities.onboard.MainActivity;
import com.shakil.barivara.adapter.RecyclerRoomListAdapter;
import com.shakil.barivara.databinding.ActivityRoomListBinding;
import com.shakil.barivara.firebasedb.FirebaseCrudHelper;
import com.shakil.barivara.model.room.Room;
import com.shakil.barivara.utils.CustomAdManager;
import com.shakil.barivara.utils.FilterManager;
import com.shakil.barivara.utils.Tools;
import com.shakil.barivara.utils.UX;

import java.util.ArrayList;

public class RoomListActivity extends AppCompatActivity {
    private ActivityRoomListBinding activityRoomListBinding;
    private ArrayList<Room> roomList;
    private FirebaseCrudHelper firebaseCrudHelper;
    private UX ux;
    private Tools tools;
    private ImageButton searchButton, refreshButton;
    private EditText searchName;
    private FilterManager filterManager;
    private CustomAdManager customAdManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityRoomListBinding = DataBindingUtil.setContentView(this, R.layout.activity_room_list);

        init();
        setSupportActionBar(activityRoomListBinding.toolBar);

        activityRoomListBinding.toolBar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(RoomListActivity.this, MainActivity.class));
            }
        });

        binUiWIthComponents();
    }

    private void init() {
        searchButton = findViewById(R.id.searchButton);
        refreshButton = findViewById(R.id.refreshButton);
        searchName = findViewById(R.id.SearchName);
        roomList = new ArrayList<>();
        ux = new UX(this);
        tools = new Tools(this);
        filterManager = new FilterManager();
        firebaseCrudHelper = new FirebaseCrudHelper(this);
        customAdManager = new CustomAdManager(this);
    }

    private void binUiWIthComponents() {
        //region for ad
        customAdManager.generateAd(activityRoomListBinding.adView);
        //endregion

        searchName.setHint(getString(R.string.search_room_name));
        //region check internet connection
        if (tools.hasConnection()) {
            setData();
        } else {
            Toast.makeText(this, getString(R.string.no_internet_title), Toast.LENGTH_SHORT).show();
        }
        //endregion

        activityRoomListBinding.mAddRoomMaster.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(RoomListActivity.this, RoomActivity.class));
            }
        });

        //region filter
        searchButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (tools.hasConnection()) {
                    if (!TextUtils.isEmpty(searchName.getText().toString())){
                        filterManager.onFilterClick(searchName.getText().toString(), roomList, new FilterManager.onFilterClick() {
                            @Override
                            public void onClick(ArrayList<Room> objects) {
                                if (objects.size() > 0) {
                                    roomList = objects;
                                    setRecyclerAdapter();
                                    Tools.hideKeyboard(RoomListActivity.this);
                                    Toast.makeText(RoomListActivity.this, getString(R.string.filterd), Toast.LENGTH_SHORT).show();
                                } else {
                                    Tools.hideKeyboard(RoomListActivity.this);
                                    activityRoomListBinding.mNoDataMessage.setVisibility(View.VISIBLE);
                                    activityRoomListBinding.mNoDataMessage.setText(R.string.no_data_message);
                                    activityRoomListBinding.mRecylerView.setVisibility(View.GONE);
                                    Toast.makeText(RoomListActivity.this, getString(R.string.no_data_message), Toast.LENGTH_SHORT).show();
                                }
                            }
                        });
                    }
                    else{
                        Toast.makeText(RoomListActivity.this, getString(R.string.enter_data_validation), Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(RoomListActivity.this, getString(R.string.no_internet_title), Toast.LENGTH_SHORT).show();
                }
            }
        });

        refreshButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (tools.hasConnection()) {
                    activityRoomListBinding.mRecylerView.setVisibility(View.VISIBLE);
                    searchName.setText("");
                    activityRoomListBinding.mNoDataMessage.setVisibility(View.GONE);
                    Tools.hideKeyboard(RoomListActivity.this);
                    setData();
                    Toast.makeText(RoomListActivity.this, getString(R.string.list_refreshed), Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(RoomListActivity.this, getString(R.string.no_internet_title), Toast.LENGTH_SHORT).show();
                }
            }
        });
        //endregion
    }

    //region set recycler data
    private void setData() {
        ux.getLoadingView();
        firebaseCrudHelper.fetchAllRoom("room", new FirebaseCrudHelper.onRoomDataFetch() {
            @Override
            public void onFetch(ArrayList<Room> objects) {
                roomList = objects;
                if (roomList.size()<=0){
                    activityRoomListBinding.mNoDataMessage.setVisibility(View.VISIBLE);
                    activityRoomListBinding.mNoDataMessage.setText(R.string.no_data_message);
                }
                setRecyclerAdapter();
                ux.removeLoadingView();
            }
        });
    }
    //endregion

    //region set recycler adapter
    private void setRecyclerAdapter(){
        RecyclerRoomListAdapter recyclerRoomListAdapter = new RecyclerRoomListAdapter(roomList);
        activityRoomListBinding.mRecylerView.setLayoutManager(new LinearLayoutManager(this));
        activityRoomListBinding.mRecylerView.setAdapter(recyclerRoomListAdapter);
        recyclerRoomListAdapter.notifyDataSetChanged();
        recyclerRoomListAdapter.setOnItemClickListener(new RecyclerRoomListAdapter.onItemClickListener() {
            @Override
            public void onItemClick(Room room) {
                startActivity(new Intent(RoomListActivity.this, RoomActivity.class).putExtra("room", room));
            }
        });
        recyclerRoomListAdapter.onEditListener(new RecyclerRoomListAdapter.onEditListener() {
            @Override
            public void onEdit(Room room) {
                startActivity(new Intent(RoomListActivity.this, RoomActivity.class).putExtra("room", room));
            }
        });
        recyclerRoomListAdapter.onDeleteListener(new RecyclerRoomListAdapter.onDeleteListener() {
            @Override
            public void onDelete(Room room) {
                doPopUpForDeleteConfirmation(room);
            }
        });
    }
    //endregion

    //region ask to delete confirmation
    private void doPopUpForDeleteConfirmation(Room room){
        Button cancel, delete;
        Dialog dialog = new Dialog(RoomListActivity.this, android.R.style.Theme_Dialog);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.delete_confirmation_layout);
        dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        dialog.setCanceledOnTouchOutside(true);

        cancel = dialog.findViewById(R.id.cancelButton);
        delete = dialog.findViewById(R.id.deleteButton);

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dialog.dismiss();
            }
        });

        delete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                firebaseCrudHelper.deleteRecord("room",room.getFireBaseKey());
                dialog.dismiss();
                setData();
            }
        });

        dialog.setCanceledOnTouchOutside(false);
        dialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
        dialog.getWindow().setLayout(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
        dialog.show();
    }
    //endregion

    @Override
    public void onBackPressed() {
        startActivity(new Intent(RoomListActivity.this,MainActivity.class));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/shakil/barivara/presentation/room/RoomListActivity.kt
package com.shakil.barivara.presentation.room

import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.Button
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.shakil.barivara.BaseActivity
import com.shakil.barivara.R
import com.shakil.barivara.data.model.room.Room
import com.shakil.barivara.databinding.ActivityRoomListBinding
import com.shakil.barivara.presentation.adapter.RecyclerRoomListAdapter
import com.shakil.barivara.presentation.adapter.RecyclerRoomListAdapter.RoomCallBacks
import com.shakil.barivara.presentation.onboard.HomeActivity
import com.shakil.barivara.utils.ButtonActionConstants
import com.shakil.barivara.utils.Constants.mAccessToken
import com.shakil.barivara.utils.PrefManager
import com.shakil.barivara.utils.ScreenNameConstants
import com.shakil.barivara.utils.Tools
import com.shakil.barivara.utils.UX
import com.shakil.barivara.utils.filterList
import com.shakil.barivara.utils.textChanges
import dagger.hilt.android.AndroidEntryPoint
import es.dmoral.toasty.Toasty
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

@AndroidEntryPoint
class RoomListActivity : BaseActivity<ActivityRoomListBinding>(), RoomCallBacks {
    private lateinit var activityRoomListBinding: ActivityRoomListBinding
    private lateinit var ux: UX
    private var tools = Tools(this)
    private lateinit var prefManager: PrefManager
    private val viewModel by viewModels<RoomViewModel>()
    private lateinit var recyclerRoomListAdapter: RecyclerRoomListAdapter
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    private val onBackPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            buttonAction(
                ButtonActionConstants.actionRoomListClose
            )
            startActivity(
                Intent(
                    this@RoomListActivity,
                    HomeActivity::class.java
                )
            )
        }
    }
    override val layoutResourceId: Int
        get() = R.layout.activity_room_list

    override fun setVariables(dataBinding: ActivityRoomListBinding) {
        activityRoomListBinding = dataBinding
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        screenViewed(ScreenNameConstants.appSreenRoomList)
        setSupportActionBar(activityRoomListBinding.toolBar)
        activityRoomListBinding.searchLayout.SearchName.hint = getString(R.string.search_room_name)

        init()
        initListeners()
        initObservers()
        setRecyclerAdapter()
        viewModel.getAllRooms(prefManager.getString(mAccessToken))
    }

    private fun init() {
        ux = UX(this)
        prefManager = PrefManager(this)
        setupDebouncedSearch()
    }

    private fun initListeners() {
        onBackPressedDispatcher.addCallback(onBackPressedCallback)
        activityRoomListBinding.toolBar.setNavigationOnClickListener {
            buttonAction(
                ButtonActionConstants.actionRoomListClose
            )
            startActivity(
                Intent(
                    this@RoomListActivity,
                    HomeActivity::class.java
                )
            )
        }

        activityRoomListBinding.mAddRoomMaster.setOnClickListener {
            buttonAction(
                ButtonActionConstants.actionAddNewRoom
            )
            startActivity(
                Intent(
                    this@RoomListActivity,
                    RoomActivity::class.java
                )
            )
        }

        activityRoomListBinding.searchLayout.refreshButton.setOnClickListener {
            if (tools.hasConnection()) {
                activityRoomListBinding.mRecylerView.visibility = View.VISIBLE
                activityRoomListBinding.searchLayout.SearchName.setText("")
                activityRoomListBinding.noDataLayout.root.visibility = View.GONE
                Tools.hideKeyboard(this@RoomListActivity)
                viewModel.getAllRooms(prefManager.getString(mAccessToken))
                Toast.makeText(
                    this@RoomListActivity,
                    getString(R.string.list_refreshed),
                    Toast.LENGTH_SHORT
                ).show()
            } else {
                Toast.makeText(
                    this@RoomListActivity,
                    getString(R.string.no_internet_title),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    private fun setupDebouncedSearch() {
        activityRoomListBinding.searchLayout.SearchName.textChanges()
            .debounce(100)
            .onEach { query ->
                val originalList = viewModel.getRooms().value ?: emptyList()
                // Use the common filter function
                val filteredList =
                    filterList(query.toString().lowercase(), originalList) { room, q ->
                        room.name.lowercase().contains(q, ignoreCase = true)
                    }

                if (filteredList.isNotEmpty()) {
                    activityRoomListBinding.mRecylerView.visibility = View.VISIBLE
                    activityRoomListBinding.noDataLayout.root.visibility =
                        View.GONE
                    recyclerRoomListAdapter.setItems(filteredList)
                    Tools.hideKeyboard(this@RoomListActivity)
                    Toast.makeText(
                        this@RoomListActivity,
                        getString(R.string.filterd),
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    Tools.hideKeyboard(this@RoomListActivity)
                    activityRoomListBinding.mRecylerView.visibility = View.GONE
                    activityRoomListBinding.noDataLayout.root.visibility =
                        View.VISIBLE
                    Toast.makeText(
                        this@RoomListActivity,
                        getString(R.string.no_data_message),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
            .launchIn(coroutineScope) // Launch in the Coroutine scope
    }

    private fun initObservers() {
        viewModel.getRooms().observe(this) { rooms ->
            if (rooms.isEmpty()) {
                activityRoomListBinding.noDataLayout.root.visibility = View.VISIBLE
                activityRoomListBinding.mRecylerView.visibility = View.GONE
            } else {
                recyclerRoomListAdapter.setItems(rooms)
                activityRoomListBinding.mRecylerView.visibility = View.VISIBLE
                activityRoomListBinding.noDataLayout.root.visibility = View.GONE
            }
        }

        viewModel.getDeleteRoomResponse().observe(this) { deleteResponse ->
            if (deleteResponse.statusCode == 200) {
                viewModel.getAllRooms(prefManager.getString(mAccessToken))
            } else {
                Toasty.warning(this, deleteResponse.message, Toast.LENGTH_SHORT).show()
            }
        }

        viewModel.isLoading.observe(this) { isLoading ->
            if (isLoading) {
                ux.getLoadingView()
            } else {
                ux.removeLoadingView()
            }
        }
    }

    private fun setRecyclerAdapter() {
        recyclerRoomListAdapter = RecyclerRoomListAdapter()
        activityRoomListBinding.mRecylerView.layoutManager = LinearLayoutManager(this)
        activityRoomListBinding.mRecylerView.adapter = recyclerRoomListAdapter
        recyclerRoomListAdapter.setRoomCallBack(this)
    }

    private fun doPopUpForDeleteConfirmation(room: Room) {
        val dialog = Dialog(this@RoomListActivity, android.R.style.Theme_Dialog)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.delete_confirmation_layout)
        dialog.window?.setBackgroundDrawableResource(android.R.color.transparent)
        dialog.setCanceledOnTouchOutside(true)
        val cancel: Button = dialog.findViewById(R.id.cancelButton)
        val delete: Button = dialog.findViewById(R.id.deleteButton)
        cancel.setOnClickListener { dialog.dismiss() }
        delete.setOnClickListener {
            viewModel.deleteRoom(prefManager.getString(mAccessToken), room.id)
            dialog.dismiss()
        }
        dialog.setCanceledOnTouchOutside(false)
        dialog.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
        dialog.window?.setLayout(
            RelativeLayout.LayoutParams.MATCH_PARENT,
            RelativeLayout.LayoutParams.WRAP_CONTENT
        )
        dialog.show()
    }

    override fun onDelete(room: Room) {
        buttonAction(
            ButtonActionConstants.actionRoomDelete
        )
        doPopUpForDeleteConfirmation(room)
    }

    override fun onEdit(room: Room) {
        buttonAction(
            ButtonActionConstants.actionRoomUpdate
        )
        startActivity(
            Intent(this@RoomListActivity, RoomActivity::class.java).putExtra(
                "room",
                room
            )
        )
    }

    override fun onItemClick(room: Room) {
        buttonAction(
            ButtonActionConstants.actionRoomUpdate
        )
        startActivity(
            Intent(this@RoomListActivity, RoomActivity::class.java).putExtra(
                "room",
                room
            )
        )
    }
}
